File: 백준/Silver/1874. 스택 수열/스택 수열.cc
#include <iostream>
#include <stack>
#include <queue>
using namespace std;

int main() {
	int n;
	int cnt = 1;
	stack <int> s;
	queue <char> res;
	cin >> n;

	for (int i = 0; i < n; i++) {
		int x;
		cin >> x;

		while (cnt <= x) {
			s.push(cnt);
			cnt++;
			res.push('+');
		}

		if (s.top() == x) {
			s.pop();
			res.push('-');
		}
		else {
			cout << "NO" << "\n";
			return 0;
		}
	}

	while (res.size() > 0) {
		cout << res.front() << "\n";
		res.pop();
	}
}

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.util.Scanner;
import java.util.Stack;

public class Main {
    public static void main(String[] args) {
        Stack <Integer> stack = new Stack<>();
        Scanner sc = new Scanner(System.in);
        StringBuilder sb = new StringBuilder();

        int top = 0;
        boolean check = true;
        int n = sc.nextInt();
        for(int i = 0; i < n; i++){
            int x = sc.nextInt();
            // 스택에 넣기 (+)
            if(x > top){
                for(int j = top + 1; j <= x; j++){
                    stack.push(j);
                    sb.append("+").append("\n");
                }
                top = x;
            }

            // 맨 위를 비교
            if(x != stack.peek()){
                check = false;
                break;
            }

            // 스택에서 빼기 (-)
            sb.append("-").append("\n");
            stack.pop();
        }

        if(check) System.out.println(sb.toString());
        else System.out.println("NO");

    }
}
